File: frontend/src/types/agreement.ts
export type AgreementStatus = 'draft' | 'active' | 'expired' | 'terminated';

export interface UserInfo {
  id: string;
  email: string;
  full_name: string | null;
}

export interface AssignmentInfo {
  id: string;
  unit_id: string;
  tenant_id: string;
  start_date: string;
  end_date: string | null;
  status: string;
}

export interface Agreement {
  id: string;
  assignment_id: string;
  landlord_id: string;
  tenant_id: string;
  start_date: string;
  end_date: string;
  monthly_rent: string;
  security_deposit: string;
  terms: string | null;
  status: AgreementStatus;
  signed_at: string | null;
  created_at: string;
  updated_at: string;
  landlord: UserInfo | null;
  tenant: UserInfo | null;
  assignment: AssignmentInfo | null;
}

export interface AgreementCreate {
  assignment_id: string;
  start_date: string;
  end_date: string;
  monthly_rent: number;
  security_deposit?: number;
  terms?: string | null;
}

export interface AgreementUpdate {
  start_date?: string;
  end_date?: string;
  monthly_rent?: number;
  security_deposit?: number;
  terms?: string | null;
}

File: backend/app/schemas/agreement.py
import uuid
from datetime import date, datetime
from typing import Optional
from decimal import Decimal
from pydantic import BaseModel, field_validator
from app.models.agreement import AgreementStatus



class UserInfo(BaseModel):
    id: uuid.UUID
    email: str
    full_name: Optional[str] = None

    @field_validator("id", mode="before")
    @classmethod
    def uuid_to_str(cls, v):
        return str(v) if isinstance(v, uuid.UUID) else v

    class Config:
        from_attributes = True


class AssignmentInfo(BaseModel):
    id: uuid.UUID
    unit_id: uuid.UUID
    tenant_id: uuid.UUID
    start_date: date
    end_date: Optional[date] = None
    status: str

    @field_validator("id", "unit_id", "tenant_id", mode="before")
    @classmethod
    def uuid_to_str(cls, v):
        return str(v) if isinstance(v, uuid.UUID) else v

    class Config:
        from_attributes = True


class AgreementCreate(BaseModel):
    assignment_id: uuid.UUID
    start_date: date
    end_date: date
    monthly_rent: Decimal
    security_deposit: Optional[Decimal] = Decimal("0")
    terms: Optional[str] = None


class AgreementUpdate(BaseModel):
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    monthly_rent: Optional[Decimal] = None
    security_deposit: Optional[Decimal] = None
    terms: Optional[str] = None


class AgreementResponse(BaseModel):
    id: uuid.UUID
    assignment_id: uuid.UUID
    landlord_id: uuid.UUID
    tenant_id: uuid.UUID
    start_date: date
    end_date: date
    monthly_rent: Decimal
    security_deposit: Decimal
    terms: Optional[str] = None
    status: AgreementStatus
    signed_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    landlord: Optional[UserInfo] = None
    tenant: Optional[UserInfo] = None
    assignment: Optional[AssignmentInfo] = None
    tenant_accepted_at: Optional[datetime] = None

    @field_validator("id", "assignment_id", "landlord_id", "tenant_id", mode="before")
    @classmethod
    def uuid_to_str(cls, v):
        return str(v) if isinstance(v, uuid.UUID) else v

    class Config:
        from_attributes = True